### PR TITLE
Add shader hash blob part

### PR DESCRIPTION
**Pull request by tex3d**
Includes minor behavior changes regarding debug name/hashing:
1. Default to hashing shader module rather than debug module, so separate compilation for debug version can match by default, without having to include special flags.
2. Produce debug name and hash when using -Zsb, even without -Zi or -Fd

### CODE REVIEW

File: tools/clang/unittests/HLSL/DxilContainerTest.cpp [hunk 1]
```diff
@@ -419,7 +422,17 @@ class DxilContainerTest : public ::testing::Test {
...
     VERIFY_SUCCEEDED(hrVer);
     VERIFY_SUCCEEDED(pVersionInfo->GetVersion(&Major, &Minor));
-    return Major == 1 && (Minor >= 1);
+    return !(Major == 1 && Minor < 1);
+  }
+
+  bool DoesValidatorSupportShaderHash() {
+    CComPtr<IDxcVersionInfo> pVersionInfo;
+    UINT Major, Minor;
+    HRESULT hrVer = m_dllSupport.CreateInstance(CLSID_DxcValidator, &pVersionInfo);
```
Comment (MrTrillian) [inline tools/clang/unittests/HLSL/DxilContainerTest.cpp]: This assumes things about the vtable layout of the object implementing IDxcValidator. I've seen it elsewhere so I guess it works, but I'd rather we played by the rules and CreateInstance an IDxcValidator and then QI IDxcVersionInfo.

File: tools/clang/unittests/HLSL/DxilContainerTest.cpp [hunk 2]
```diff
@@ -440,6 +453,28 @@ class DxilContainerTest : public ::testing::Test {
...
+    VERIFY_SUCCEEDED(m_dllSupport.CreateInstance(CLSID_DxcContainerReflection, &pContainer));
+    VERIFY_SUCCEEDED(pContainer->Load(pProgram));
+    if (FAILED(pContainer->FindFirstPartKind(hlsl::DFCC_ShaderHash, &index))) {
+      return std::string();
+    }
+    VERIFY_SUCCEEDED(pContainer->GetPartContent(index, &pHashBlob));
+    const hlsl::DxilShaderHash *pShaderHash = (hlsl::DxilShaderHash *)pHashBlob->GetBufferPointer();
+    std::string result;
+    llvm::raw_string_ostream os(result);
+    for (int i = 0; i < 16; ++i)
```
Comment (MrTrillian) [inline tools/clang/unittests/HLSL/DxilContainerTest.cpp]: Don't block the checkin on this but `for (char b : pShaderHash->Digest)` would save you from hard-coding `16`. There's another such loop in this CR.

File: tools/clang/unittests/HLSL/DxilContainerTest.cpp (hunk 1 above)
Comment (tex3d) [reply tools/clang/unittests/HLSL/DxilContainerTest.cpp]: Actually CreateInstance does the QI based on the passed in interface, doesn't it?

File: tools/clang/unittests/HLSL/DxilContainerTest.cpp (hunk 2 above)
Comment (tex3d) [reply tools/clang/unittests/HLSL/DxilContainerTest.cpp]: Thanks, I was quickly copy-pasting it from the function that converted the digest to hex inside the md5 code.  But good suggestion.

File: tools/clang/unittests/HLSL/DxilContainerTest.cpp (hunk 1 above)
Comment (tex3d) [reply tools/clang/unittests/HLSL/DxilContainerTest.cpp]: Yeah, each Create* implementation for each object type returns result of QI on interface.

**Pull request merged**